### DISCUSSION TITLE
Should namespaces be globally defined?

### DISCUSSION THREAD

**Discussion by jhiemstrawisc** [PelicanPlatform/pelican#920] [General]
As we start to work on disentangling origins and namespaces, we've run into the issue where multiple origins may export the same namespace prefix with different advertisements. Under the hood we map an origin advertisement to the list of namespace advertisements it exports, where those namespace advertisements are defined by the origin. So, for example, if `origin1` and `origin2` both export the namespace prefix `/foo`, they may do so such that `origin1` says `/foo` is publicly readable and `origin2` says it isn't.

This begs the question -- should these namespace advertisements be global, preventing two origins from trying to represent the same namespace prefix differently? Currently we are not equipped to handle the case where two origins have misaligned namespace ads -- we simply grab the namespace ad from whichever origin is the first we match against. Using the previous example, this means that if we match against `origin1` first, our headers will indicate the namespace is public and thus provide no token generation information. However, the client might still try to read from the namespace prefix at `origin2` if it makes use of the metalink headers.

**Comment by jhiemstrawisc** [comment]
@bbockelm this is something we tried to discuss during the team meeting on 03/08/24 but quickly decided we needed your input.

**Comment by joereuss12** [comment]
This also ties into what should the `collections-url` point at? Currently, the url points to the origin and is populated both if `dirlisthost` is set or if the export enables listings. This gets confusing very quickly if there is 2 origins are exporting the same namespace. What origin does the collections-url point to? I'm pretty sure as of now, it points to whichever origin we find first.

In addition, within the client code, we changed looking for `dirlisthost` to instead do a `PROPFIND` request against the director. This `PROPFIND` request tells the director to immediately redirect to an origin and give us the first origin it finds with that prefix (that enables listings). We thought a better solution might be to have the director redirect to whatever the collections-url is set to, that way if we have multiple origins all exporting the same namespace, we can define which origin we want to be the collections-url for that namespace.

There were a few other ideas that came up when discussing this with @jhiemstrawisc but this is becoming a little confusing. I think we need to have a discussion about this with the whole team
